### PR TITLE
Support bfloat16 and complex dtypes for logical_not

### PR DESCRIPTION
**Pull request by xuhdev**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#43537 Support bfloat16 and complex dtypes for logical_not**
* #43174 Add complex number support for binary logical operators

Differential Revision: [D23751950](https://our.internmc.facebook.com/intern/diff/D23751950)

### CODE REVIEW

File: test/test_torch.py [hunk 1]
```diff
@@ -6266,39 +6266,32 @@ def test_bitwise_xor(self, device):
...
-        # bfloat16 tensors.  Can we define our own dtype in NumPy so testing would be easier?
-        if dtype == torch.bfloat16 or dtype.is_complex:
-            self.assertRaises(RuntimeError, lambda: a.logical_not())
-            self.assertRaises(RuntimeError, lambda: a.logical_not_())
-            raise unittest.SkipTest('logical_not not supported on {}'.format(dtype))
-
-        a_np = np.array(data, dtype=torch_to_numpy_dtype_dict[dtype])
-        self.assertEqual(np.logical_not(a_np), torch.logical_not(a).to('cpu'))
-        self.assertEqual(np.logical_not(a_np, out=a_np), a.logical_not_().to('cpu'))
+        if dtype == torch.bfloat16 or dtype.is_complex:  # numpy doesn't support these dtypes
```
Comment (mruberry) [inline test/test_torch.py:6269]: NumPy does support logical not on complex types, doesn't it? 

```
a = np.array((1 + 1j))
np.logical_not(a)
: False
```

File: test/test_torch.py [hunk 2]
```diff
@@ -6266,39 +6266,32 @@ def test_bitwise_xor(self, device):
...
-            raise unittest.SkipTest('logical_not not supported on {}'.format(out_dtype))
-
-        out_np = np.empty(a.shape, dtype=torch_to_numpy_dtype_dict[out_dtype])
-
-        self.assertEqual(a, a.cpu().numpy())
-        torch.logical_not(a, out=out)
-        np.logical_not(a.cpu().numpy(), out=out_np)
-        self.assertEqual(out_np, out.to('cpu'))
+        out = torch.empty_like(a, dtype=out_dtype, device=device)
+        if torch.bfloat16 in dtypes or dtype.is_complex or out_dtype.is_complex:  # numpy doesn't support these dtypes
```
Comment (mruberry) [inline test/test_torch.py:6286]: Same question here about complex NumPy support

File: test/test_torch.py (hunk 1 above)
Comment (mruberry) [reply test/test_torch.py:6269]: For bfloat16 you can set an explicit dtype of float32 with a pattern like:

```
if dtype is torch.bfloat16:
  np_dtype = np.float32
else:
  np_dtype = torch_to_numpy_dtype_dict[dtype]
```

**Comment by mruberry** [comment]
Hey @xuhdev, sorry this went so long without a response. I've approved the other 3 PRs in this stack but have a question about this one.

File: test/test_torch.py (hunk 1 above)
Comment (xuhdev) [reply test/test_torch.py:6269]: Sorry for the late reply. I've removed complex from this line.

I've tried your suggestion but `self.assertEqual(a, a.cpu().numpy())` (a few lines below) always leads to `TypeError` for bfloat16.

I have also rebased the stack.

**Review by mruberry — APPROVED** [review]
Nice cleanup! Thanks @xuhdev. Now I have to get this stack merged.

**Comment by mruberry** [comment]
Congrats, @xuhdev! Finally the entire stack is landed!

**Comment by xuhdev** [comment]
@mruberry Finally! Thanks for all the landing work!

**Pull request closed without merging**